cli/commands/scaffold/action.go

```diff
@@ -0,0 +1,351 @@
+	parsedModuleUrl, err = addRefToModuleUrl(opts, parsedModuleUrl, vars)
+	if err != nil {
+		return errors.WithStackTrace(err)
+	}
+
+	// regenerate module url with all changes
+	moduleUrl = parsedModuleUrl.String()
+
+	// identify template url
+	templateDir := ""
+	if templateUrl != "" {
+		// process template url if was passed
+		parsedTemplateUrl, err := terraform.ToSourceUrl(templateUrl, tempDir)
+		if err != nil {
+			return errors.WithStackTrace(err)
+		}
+		parsedTemplateUrl, err = rewriteTemplateUrl(opts, parsedTemplateUrl)
+		if err != nil {
+			return errors.WithStackTrace(err)
+		}
+		// regenerate template url with all changes
+		templateUrl = parsedTemplateUrl.String()
+
+		templateDir, err = os.MkdirTemp("", "template")
+		if err != nil {
+			return errors.WithStackTrace(err)
+		}
+		dirsToClean = append(dirsToClean, templateDir)
+
+		// downloading template
+		opts.Logger.Infof("Using template from %s", templateUrl)
+		if err := getter.GetAny(templateDir, templateUrl); err != nil {
+			return errors.WithStackTrace(err)
+		}
+	}
+
+	opts.Logger.Infof("Scaffolding a new Terragrunt module %s %s to %s", moduleUrl, templateUrl, opts.WorkingDir)
+	if err := getter.GetAny(tempDir, moduleUrl); err != nil {
+		return errors.WithStackTrace(err)
+	}
+
+	// extract variables from module url
+	inputs, err := config.ParseVariables(opts, tempDir)
+	if err != nil {
+		return errors.WithStackTrace(err)
+	}
+
+	// separate variables that require value and with default value
+	var requiredVariables []*config.ParsedVariable
+	var optionalVariables []*config.ParsedVariable
+
+	for _, value := range inputs {
+		if value.DefaultValue == "" {
+			requiredVariables = append(requiredVariables, value)
+		} else {
+			optionalVariables = append(optionalVariables, value)
+		}
+	}
+	opts.Logger.Debugf("Parsed %d required variables and %d optional variables", len(requiredVariables), len(optionalVariables))
+
+	// run boilerplate
+
+	// prepare boilerplate dir
+	boilerplateDir := util.JoinPath(tempDir, util.DefaultBoilerplateDir)
+	// use template dir as boilerplate dir
+	if templateDir != "" {
+		boilerplateDir = templateDir
+	}
+
+	// if boilerplate dir is not found, create one with default template
+	if !files.IsExistingDir(boilerplateDir) {
+		// no default boilerplate dir, create one
+		boilerplateDir, err = os.MkdirTemp("", "boilerplate")
+		if err != nil {
+			return errors.WithStackTrace(err)
+		}
+		dirsToClean = append(dirsToClean, boilerplateDir)
+		if err := os.WriteFile(util.JoinPath(boilerplateDir, "terragrunt.hcl"), []byte(defaultTerragruntTemplate), 0644); err != nil {
+			return errors.WithStackTrace(err)
+		}
+		if err := os.WriteFile(util.JoinPath(boilerplateDir, "boilerplate.yml"), []byte(defaultBoilerplateConfig), 0644); err != nil {
+			return errors.WithStackTrace(err)
+		}
+	}
+
+	// add additional variables
+	vars["requiredVariables"] = requiredVariables
+	vars["optionalVariables"] = optionalVariables
+
+	vars["sourceUrl"] = moduleUrl
```

Comment:
We need to treat this and document it as an API. 

That is, whatever variables Terragrunt makes available to the boilerplate template need to be well-defined and captured somewhere as an explicit public API, and one for which we'll want to maintain backward compatibility (and document breaking changes & migration guides if necessary). Consider defining these as a struct or somewhere centralized, with appropriate comments, so it's clear this is an API that we need to treat carefully.

Reply:
Bumping this one more time: do we have a test that checks for this API? I'm worried we'll break it in a backward incompatible way by accident, and not notice... And any custom templates users created will suddenly no longer work.